File: examples/dekita/.claude/hooks/session_metrics_collector.py
#!/usr/bin/env python3
"""
セッションメトリクス収集フック（Stop）

セッション終了時に自動でメトリクスを収集・記録する。
"""

# SRP: セッション終了時のメトリクス収集のみを担当（単一責任）
# 既存フックとの重複なし（新規機能）
# ブロックなし（情報収集のみのため）

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

# 共通モジュール
HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))
from lib.constants import TIMEOUT_HEAVY
from lib.execution import log_hook_execution
from lib.logging import get_error_context_manager
from lib.session import create_hook_context, parse_hook_input

SCRIPT_DIR = HOOKS_DIR.parent / "scripts"


def collect_session_metrics(session_id: str) -> bool:
    """セッションメトリクスを収集

    Args:
        session_id: Claude Codeから提供されたセッションID
    """
    collect_script = SCRIPT_DIR / "collect-session-metrics.py"
    if not collect_script.exists():
        return False

    try:
        # Issue #2317: 環境変数ではなくコマンドライン引数でsession_idを渡す
        result = subprocess.run(
            ["python3", str(collect_script), "--session-id", session_id],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_HEAVY,
        )
        return result.returncode == 0
    except Exception:
        return False


def generate_session_report(session_id: str) -> bool:
    """セッションレポートを生成

    Issue #1367: セッション終了時に統合レポートを生成

    Args:
        session_id: Claude Codeから提供されたセッションID
    """
    report_script = SCRIPT_DIR / "session-report-generator.py"
    if not report_script.exists():
        return False

    try:
        # Issue #2317: 環境変数ではなくコマンドライン引数でsession_idを渡す
        result = subprocess.run(
            ["python3", str(report_script), "--session-id", session_id],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_HEAVY,
        )
        return result.returncode == 0
    except Exception:
        return False


def main():
    """Collect session metrics on session end."""
    # Stop hookはstdinからJSON入力を受け取る
    hook_input = parse_hook_input()

    ctx = create_hook_context(hook_input)

    # Stop hookが既にアクティブな場合は即座にapprove
    if hook_input.get("stop_hook_active"):
        print(json.dumps({"decision": "approve"}))
        return

    # セッションIDを取得（Issue #1308: hook入力から直接取得を優先）
    # hook入力にsession_idがある場合はそれを使用し、
    # ない場合（None、空文字列含む）のみctx.get_session_id()にフォールバック
    # 注: 空文字列も無効なsession_idとして扱い、フォールバックする（意図的な動作）
    session_id = hook_input.get("session_id") or ctx.get_session_id()

    # エラーコンテキストをフラッシュ（Issue #1636）
    # セッション終了時に pending のエラーコンテキストを保存
    error_context_manager = get_error_context_manager()
    context_flushed = error_context_manager.flush_pending(session_id) is not None

    # メトリクス収集（非ブロッキング）
    metrics_success = collect_session_metrics(session_id)

    # レポート生成（Issue #1367）
    report_success = generate_session_report(session_id)

    log_hook_execution(
        "session_metrics_collector",
        "approve",
        f"Session metrics {'collected' if metrics_success else 'collection failed'}, "
        f"report {'generated' if report_success else 'generation failed'}, "
        f"error context {'flushed' if context_flushed else 'no pending'}",
        {
            "metrics_success": metrics_success,
            "report_success": report_success,
            "context_flushed": context_flushed,
        },
    )

    # メトリクス収集・レポート生成の成否に関わらずapprove
    print(json.dumps({"decision": "approve"}))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/session_metrics_collector.ts
#!/usr/bin/env bun
/**
 * セッションメトリクス収集フック（Stop）
 *
 * セッション終了時に自動でメトリクスを収集・記録する。
 *
 * Why:
 *   セッションの品質を継続的に改善するため、終了時にメトリクスを
 *   収集して分析可能にする必要がある。
 *
 * What:
 *   - collect_session_metrics.pyでメトリクス収集
 *   - session_report_generator.pyでレポート生成
 *   - ブロックせず、情報収集のみ
 *
 * Remarks:
 *   - SRP: セッション終了時のメトリクス収集のみを担当
 *   - 既存フックとの重複なし
 *   - ブロックなし（情報収集のみのため）
 *   - Python版: session_metrics_collector.py からの移行
 *
 * Changelog:
 *   - silenvx/dekita#1367: セッションレポート生成追加
 *   - silenvx/dekita#1636: エラーコンテキストフラッシュ追加
 *   - silenvx/dekita#2317: session_idをコマンドライン引数で渡すように変更
 *   - silenvx/dekita#3142: TypeScriptに移植
 */

import { spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { TIMEOUT_HEAVY } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { getErrorContextManager, logHookExecution } from "../lib/logging";
import { parseHookInput } from "../lib/session";
import { createHookContext } from "../lib/types";

const HOOK_NAME = "session-metrics-collector";

/** Scripts directory */
const SCRIPTS_DIR = join(import.meta.dir, "..", "..", "scripts");

/**
 * Collect session metrics.
 *
 * @param sessionId - Claude Code session ID
 * @returns true if collection succeeded
 */
export function collectSessionMetrics(sessionId: string): boolean {
  const collectScript = join(SCRIPTS_DIR, "collect_session_metrics.py");

  if (!existsSync(collectScript)) {
    return false;
  }

  try {
    // Issue #2317: Pass session_id via command line argument instead of env
    const result = spawnSync("python3", [collectScript, "--session-id", sessionId], {
      encoding: "utf-8",
      timeout: TIMEOUT_HEAVY * 1000, // Convert to ms
    });
    return result.status === 0;
  } catch {
    return false;
  }
}

/**
 * Generate session report.
 *
 * Issue #1367: Generate integrated report at session end
 *
 * @param sessionId - Claude Code session ID
 * @returns true if generation succeeded
 */
export function generateSessionReport(sessionId: string): boolean {
  const reportScript = join(SCRIPTS_DIR, "session_report_generator.py");

  if (!existsSync(reportScript)) {
    return false;
  }

  try {
    // Issue #2317: Pass session_id via command line argument instead of env
    const result = spawnSync("python3", [reportScript, "--session-id", sessionId], {
      encoding: "utf-8",
      timeout: TIMEOUT_HEAVY * 1000, // Convert to ms
    });
    return result.status === 0;
  } catch {
    return false;
  }
}

async function main(): Promise<void> {
  // Stop hook receives JSON input from stdin
  const hookInput = await parseHookInput();

  // If stop hook is already active, approve immediately
  if (hookInput.stop_hook_active) {
    console.log(JSON.stringify({}));
    return;
  }

  // Get session ID (prefer hook input, fallback to context)
  // Issue #1308: Hook input takes priority, fallback to ctx.sessionId
  // Note: Empty string is treated as invalid and triggers fallback (intentional)
  const ctx = createHookContext(hookInput);
  const sessionId = hookInput.session_id || ctx.sessionId || "unknown";

  // Flush pending error context (Issue #1636)
  // Save any pending error context at session end
  const errorContextManager = getErrorContextManager();
  const contextFlushed = (await errorContextManager.flushPending(sessionId)) !== undefined;

  // Metrics collection (non-blocking)
  const metricsSuccess = collectSessionMetrics(sessionId);

  // Report generation (Issue #1367)
  const reportSuccess = generateSessionReport(sessionId);

  logHookExecution(
    HOOK_NAME,
    "approve",
    `Session metrics ${metricsSuccess ? "collected" : "collection failed"}, ` +
      `report ${reportSuccess ? "generated" : "generation failed"}, ` +
      `error context ${contextFlushed ? "flushed" : "no pending"}`,
    {
      metrics_success: metricsSuccess,
      report_success: reportSuccess,
      context_flushed: contextFlushed,
    },
  );

  // Always approve regardless of collection/report success
  console.log(JSON.stringify({}));
}

// Execute
if (import.meta.main) {
  main().catch((error) => {
    console.error(`[${HOOK_NAME}] Error: ${formatError(error)}`);
    // Still approve on error to not block session end
    console.log(JSON.stringify({}));
  });
}
